### PR TITLE
CR-1090645 Be able to reinstate revert to golden to work only when it…

### PR DESCRIPTION
**Pull request by xuhz**
…'s the correct card
xbmgmt2 test example,

case 1: there is no golden file in base pkg,
root@xsjbrd100bx:/opt/xilinx/firmware/u30/gen3x4/base/data# /tmp/xbmgmt2 program --revert-to-golden -d 17:00.0
INFO     : Resetting card [ 0000:17:00.0 ] back to factory mode. 
Are you sure you wish to proceed? [Y/n]: y
ERROR: Doesn't find golden in base pkg

case2: the golden in base pkg is not the same to the one on flash
root@xsjbrd100bx:/opt/xilinx/firmware/u30/gen3x4/base/data# /tmp/xbmgmt2 program --revert-to-golden -d 17:00.0
INFO     : Resetting card [ 0000:17:00.0 ] back to factory mode. 
Are you sure you wish to proceed? [Y/n]: y
[FAILED] : Flash verified < 0s >
[ERROR]: Doesn't find valid golden on flash !!

case3: normal case flash one device
root@xsjbrd100bx:/opt/xilinx/firmware/u30/gen3x4/base/data# /tmp/xbmgmt2 program --revert-to-golden -d 17:00.0
INFO     : Resetting card [ 0000:17:00.0 ] back to factory mode. 
Are you sure you wish to proceed? [Y/n]: y
[PASSED] : Flash verified < 5s >
[PASSED] : Flash erased < 1m 49s >
INFO     : Shell on [ 0000:17:00.0 ] is reset successfully.
****************************************************
Cold reboot machine to load the new image on card(s).
****************************************************

case4: normal case flash all devices
root@xsjbrd100bx:/opt/xilinx/firmware/u30/gen3x4/base/data# /tmp/xbmgmt2 program --revert-to-golden -d all
INFO     : Resetting card [ 0000:66:00.0 ] back to factory mode. 
INFO     : Resetting card [ 0000:18:00.0 ] back to factory mode. 
INFO     : Resetting card [ 0000:17:00.0 ] back to factory mode. 
Are you sure you wish to proceed? [Y/n]: n
Action canceled.
root@xsjbrd100bx:/opt/xilinx/firmware/u30/gen3x4/base/data# echo 1 >/sys/bus/pci/devices/0000\:66\:00.1/remove
root@xsjbrd100bx:/opt/xilinx/firmware/u30/gen3x4/base/data# echo 1 >/sys/bus/pci/devices/0000\:66\:00.0/remove
root@xsjbrd100bx:/opt/xilinx/firmware/u30/gen3x4/base/data# /tmp/xbmgmt2 program --revert-to-golden -d all
INFO     : Resetting card [ 0000:18:00.0 ] back to factory mode. 
INFO     : Resetting card [ 0000:17:00.0 ] back to factory mode. 
Are you sure you wish to proceed? [Y/n]: y
[PASSED] : Flash verified < 5s >
[PASSED] : Flash erased < 1m 47s >
INFO     : Shell on [ 0000:18:00.0 ] is reset successfully.
[PASSED] : Flash verified < 5s >
[PASSED] : Flash erased < 7s >
INFO     : Shell on [ 0000:17:00.0 ] is reset successfully.
****************************************************
Cold reboot machine to load the new image on card(s).
****************************************************

### CODE REVIEW

**Review by mamin506 — APPROVED** [review]

File: src/runtime_src/core/tools/xbmgmt2/flash/flasher.cpp [hunk 1]
```diff
@@ -124,7 +126,14 @@ int Flasher::upgradeFirmware(const std::string& flasherType,
         XQSPIPS_Flasher xqspi_ps(m_device);
         if (primary == nullptr)
         {
-            retVal = xqspi_ps.revertToMFG();
+            std::string golden_file = getQspiGolden();
+            if (golden_file.empty()) {
+                std::cout << "ERROR: Doesn't find golden in base pkg" << std::endl;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/flash/flasher.cpp:131]: Change error message to "Golden image not found in the base package. Can't revert to golden"

File: src/runtime_src/core/tools/xbmgmt2/flash/xqspips.cpp [hunk 2]
```diff
@@ -535,25 +480,113 @@ int XQSPIPS_Flasher::xclUpgradeFirmware(std::istream& binStream)
...
+
+int XQSPIPS_Flasher::revertToMFG(std::istream& binStream)
+{
+    initQSpiPS();
+
+    uint32_t StatusReg = XQSpiPS_GetStatusReg();
+
+    if (StatusReg == 0xFFFFFFFF) {
+        std::cout << "[ERROR]: Read PCIe device return -1. Cannot get QSPI status." << std::endl;
+        exit(-EOPNOTSUPP);
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/flash/xqspips.cpp:517]: should this be exit or return?

**Review by AShivangi — COMMENTED** [review]
Minor comments

File: src/runtime_src/core/tools/xbmgmt2/flash/flasher.cpp (hunk 1 above)
Comment (xuhz) [reply src/runtime_src/core/tools/xbmgmt2/flash/flasher.cpp]: changed.

File: src/runtime_src/core/tools/xbmgmt2/flash/xqspips.cpp (hunk 2 above)
Comment (xuhz) [reply src/runtime_src/core/tools/xbmgmt2/flash/xqspips.cpp:517]: this kind of error is not xclmgmt can handle. Just exit

**Review by AShivangi — APPROVED** [review]

**Pull request merged**